DSA-Campaign-Uplift-Estimation/src/main/java/com/google/sps/classes/DSACampaign.java

```diff
@@ -69,8 +69,10 @@
     public int clicks;
     // total money spent on the campaign on average in a single day
     public double cost;
+    // contains the likely queries that would be made for a website and the associated urls
+    public String[][] SQR;
 
-    public DSACampaign(String DSACampaignId, String userId, String keywordCampaignId, String name, String campaignStatus, String startDate, String endDate, double manualCPC, double dailyBudget, String locations, String negativeLocations, String domain, String targets, String adText, int impressions, int clicks, double cost) {
+    public DSACampaign(String DSACampaignId, String userId, String keywordCampaignId, String name, String campaignStatus, String startDate, String endDate, double manualCPC, double dailyBudget, String locations, String negativeLocations, String domain, String targets, String adText, int impressions, int clicks, double cost, String[][] SQR) {
```

Comment:
Is there any way we can make this a class or an array list of a class for better readability? I'm assuming you're storing as a string in datastore and not the 2-d array.

Reply:
In datastore I'm storing it as an embedded entity (the embedded entity was super useful, it allows me to store the SQR entity as a property of its associated DSA campaign entity). As far as making it a class or an array list of a class, using a String[][] is the easiest way I thought of to store the table information, and it makes displaying the data on the front end super easy. I thought it made sense to store it in a String[][] because the SQR is just a 2 column table of elements. For readability, since it only has two columns (queries and urls), it shouldn't be too difficult to read?

Reply:
In terms of readability, it would be better to have a class (ex. CampaignSQR) that contains multiple query-url pairs (ex. each query-url can be a class QueryUrlPair). This way the SQR argument is clear from the type itself. Also it is much more understandable to work with compared to indexing. Also if we wanted to add more fields such as number of impressions associated with each query-url, it would be more confusing to do so.

But I'll leave it as optional for now.